synapse/types.py

```diff
@@ -466,12 +467,12 @@ class RoomStreamToken:
     stream = attr.ib(type=int, validator=attr.validators.instance_of(int))
 
     instance_map = attr.ib(
-        type=Dict[str, int],
-        factory=dict,
+        type=Mapping[str, int],
```

Comment:
I was about to say 'huh, I never thought of just calling frozendict a Mapping!', but I'm not sure this type annotation tells the whole story.

`dict` is a subclass of `Mapping`, yet this field very clearly needs the `Mapping` to be hashable, so the type annotation is too loose here.

I think this really does need `frozendict[str, int]` — which I think you have to put in double quotes (:sob: — there is a fork of frozendict which came in after generics emerged and fixes that, but we don't use it in Synapse).

Reply:
> I think this really does need `frozendict[str, int]` — which I think you have to put in double quotes (😭 — there is a fork of frozendict which came in after generics emerged and fixes that, but we don't use it in Synapse).

We have a stub for this in our repo.

Reply:
I've updated the type to be `"frozendict[str, int]"` with the quotes. It does feel slightly off since the `frozendict`-ness is really an implementation detail. The type required is `Mapping[str, int] & Hashable` except there's no way to express that in Python that I know of.

Reply:
> > I think this really does need `frozendict[str, int]` — which I think you have to put in double quotes (sob — there is a fork of frozendict which came in after generics emerged and fixes that, but we don't use it in Synapse).
> 
> We have a stub for this in our repo.

I don't think stubs are used at runtime, so you still have to put in the quotes otherwise you get the type subscripting error. (am I missing something? I hope so :))

Reply:
> I don't think stubs are used at runtime, so you still have to put in the quotes otherwise you get the type subscripting error. (am I missing something? I hope so :))

Yep, without the quotes I get a type subscripting error.